### PR TITLE
Use spack-stack modulefiles

### LINKED ISSUE #39: Move to spack-stack
The `ufs-weather-model` has moved to spack-stack.  All of the modules need to be updated to use this mimicking what is in the ufs-srweather-app https://github.com/ufs-community/ufs-srweather-app

We should also add hercules module file here.

### PR DESCRIPTION
**Pull request by zmoon**
* [x] manually test the build on all platforms  
  e.g.
  ```
  rm -rf build && cmake -B build && cmake --build build -j 4 && mpirun -n 1 ./build/bin/nexus -h
  ```
  * [x] Hera Intel
  * [x] WCOSS
  * [x] Orion
  * [x] Gaea C5 (with the two adjustments)
  * [x] Derecho (note doesn't work on Casper)
* [x] revise the readme instructions

Closes #39

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,11 +1,20 @@
...
 
+# set compiler
+if(NOT DEFINED ENV{FC})
+  message(FATAL_ERROR "CMAKE_Fortran_COMPILER is not defined")
+endif()
+
+set(CMAKE_Fortran_COMPILER $ENV{FC})
+set(CMAKE_C_COMPILER $ENV{CC})
+set(CMAKE_CXX_COMPILER $ENV{CXX})
+
```
Comment (zmoon) [inline CMakeLists.txt]: This change is a reversion. CMake recognizes those env vars (`FC`, `CC`, etc.) automatically.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,11 +1,16 @@
 # Check environment
 cmake_minimum_required(VERSION 3.5)
 
-# Set the project name and version
+# set the project name and version
```
Comment (zmoon) [inline CMakeLists.txt]:
```suggestion
# Set the project name and version
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,11 +1,16 @@
...
 project(NEXUS
         VERSION 0.2.0
         LANGUAGES Fortran)
 
+# set compiler
+if(NOT DEFINED ENV{FC})
+  message(FATAL_ERROR "CMAKE_Fortran_COMPILER is not defined")
+endif()
```
Comment (zmoon) [inline CMakeLists.txt]: It doesn't need to be fatal error. I was experimenting with these warning messages, but the behavior seems to be not always as expected, due to cmake's caching and such I guess.
```suggestion
# FC should be used to set compiler
if(NOT DEFINED ENV{FC})
  message(WARNING "FC is not set. You may want to use it to specify a specific compiler.")
else()
  if("$ENV{FC}" STREQUAL "")
    message(WARNING "FC is set to empty string. You may want to use it to specify a specific compiler.")
  else()
    message(STATUS "FC is set: '${FC}'")
  endif()
endif()
```

File: modulefiles/ufs_common.lua [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  {["bacio"]           = "2.4.1"},
+  {["crtm"]            = "2.4.0"},
+  {["g2"]              = "3.5.1"},
+  {["g2tmpl"]          = "1.13.0"},
+  {["ip"]              = "4.3.0"},
+  {["sp"]              = "2.5.0"},
+  {["w3emc"]           = "2.10.0"},
+  {["gftl-shared"]     = "1.6.1"},
+  {["mapl"]            = "2.40.3-esmf-8.6.0"},
+  {["scotch"]          = "7.0.4"},
```
Comment (zmoon) [inline modulefiles/ufs_common.lua]: I would guess we could remove these?

Comment (bbakernoaa) [reply modulefiles/ufs_common.lua]: We probably could. At least some of them. Definitely don’t need many of these like scotch and w3emc. These are for the wave model

**Comment by bbakernoaa** [comment]
@zmoon Compiled fine on wcoss 

Did get a warning during the initial CMake step 

```bash
CMake Warning (dev) at src/CMakeLists.txt:16 (find_package):
  Policy CMP0074 is not set: find_package uses <PackageName>_ROOT variables.
  Run "cmake --help-policy CMP0074" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  Environment variable ESMF_ROOT is set to:

    /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/esmf/8.6.0

  For compatibility, CMake is ignoring the variable.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found ESMF: /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/esmf/8.6.0/lib/libesmf.a (found version "8.6.0")
CMake Warning (dev) at src/CMakeLists.txt:17 (find_package):
  Policy CMP0074 is not set: find_package uses <PackageName>_ROOT variables.
  Run "cmake --help-policy CMP0074" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  Environment variable NetCDF_ROOT is set to:

    /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2

  For compatibility, CMake is ignoring the variable.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found NetCDF: /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2/include (found version "4.9.2") found components: Fortran
-- FindNetCDF defines targets:
--   - NetCDF_VERSION [4.9.2]
--   - NetCDF_PARALLEL [TRUE]
--   - NetCDF_C_CONFIG_EXECUTABLE [/apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2/bin/nc-config]
--   - NetCDF::NetCDF_C [STATIC] [Root: /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2] Lib: /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2/lib/libnetcdf.a
--   - NetCDF_Fortran_CONFIG_EXECUTABLE [/apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2/bin/nf-config]
--   - NetCDF::NetCDF_Fortran [STATIC] [Root: /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2] Lib: /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2/lib/libnetcdff.a
CMake Warning (dev) at src/nuopc/CMakeLists.txt:16 (find_package):
  Policy CMP0074 is not set: find_package uses <PackageName>_ROOT variables.
  Run "cmake --help-policy CMP0074" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  Environment variable ESMF_ROOT is set to:

    /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/esmf/8.6.0

  For compatibility, CMake is ignoring the variable.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at src/nuopc/CMakeLists.txt:17 (find_package):
  Policy CMP0074 is not set: find_package uses <PackageName>_ROOT variables.
  Run "cmake --help-policy CMP0074" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  Environment variable NetCDF_ROOT is set to:

    /apps/prod/hpc-stack/i-19.1.3.304__m-8.1.12__h-1.14.0__n-4.9.2__p-2.5.10__e-8.6.0pnetcdf/intel-19.1.3.304/cray-mpich-8.1.12/netcdf/4.9.2

  For compatibility, CMake is ignoring the variable.
This warning is for project developers.  Use -Wno-dev to suppress it.
```

**Comment by zmoon** [comment]
We could explicitly set the policy as it suggests. For example
```cmake
if(POLICY CMP0074)
  cmake_policy(SET CMP0074 OLD)
endif()
```
I don't see UWM doing something like that in their repo though.

File: modulefiles/ufs_gaea.intel.lua [hunk 5]
```diff
@@ -0,0 +1,33 @@
+help([[
+  This module loads libraries required for building and running UFS Weather Model
+  on the NOAA RDHPC machine Gaea C5 using Intel-2023.1.0.
+]])
+
+whatis([===[Loads libraries needed for building the UFS Weather Model on Gaea ]===])
+
+prepend_path("MODULEPATH", "/ncrc/proj/epic/spack-stack/spack-stack-1.6.0/envs/fms-2024.01/install/modulefiles/Core")
+
+stack_intel_ver=os.getenv("stack_intel_ver") or "2023.1.0"
```
Comment (zmoon) [inline modulefiles/ufs_gaea.intel.lua]: Tried C5 and C6 and it seems like only 2023.2.0 is available.
```suggestion
stack_intel_ver=os.getenv("stack_intel_ver") or "2023.2.0"
```

File: modulefiles/ufs_gaea.intel.lua [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+load(pathJoin("stack-intel", stack_intel_ver))
+
+stack_cray_mpich_ver=os.getenv("stack_cray_mpich_ver") or "8.1.25"
```
Comment (zmoon) [inline modulefiles/ufs_gaea.intel.lua]: Again seems like this is the only version.
```suggestion
stack_cray_mpich_ver=os.getenv("stack_cray_mpich_ver") or "8.1.28"
```

**Comment by zmoon** [comment]
@bbakernoaa does revised readme look okay to you?

File: README.md [hunk 7]
```diff
@@ -52,58 +57,10 @@ Input data:
...
-
-# For NEXUS
-export CMAKE_Fortran_COMPILER=$ESMF_F90COMPILER
-```
-
-To build NEXUS, at least these env vars should be set:
-- `ESMFMKFILE`
-- `OMPI_FC`
-- `CMAKE_Fortran_COMPILER`
+See https://github.com/zmoon/gha-esmf for examples of how to build ESMF,
```
Comment (zmoon) [inline README.md]: Migrated
```suggestion
See https://github.com/noaa-oar-arl/gha-esmf for examples of how to build ESMF,
```

File: modulefiles/ufs_gaea.intel.lua (hunk 5 above)
Comment (zmoon) [reply modulefiles/ufs_gaea.intel.lua]: https://github.com/ufs-community/ufs-weather-model/commit/db1781a05dce1125cfe17f8324650674640f0a9e

**Comment by bbakernoaa** [comment]
whats left for this?

**Comment by zmoon** [comment]
> whats left for this?

Were you able to try WCOSS from fresh again after the change? see if works and without warnings?

**Comment by bbakernoaa** [comment]
compiled

**Comment by zmoon** [comment]
> compiled

And you didn't get those warnings from earlier in this thread?

**Comment by bbakernoaa** [comment]
No warning. Compiled just fine


Barry Baker

National Oceanic and Atmospheric Administration
Air Resources Laboratory
Physical Research Scientist
Chemical Modeling and Emissions Group Leader
NCWCP, R/ARL, Rm. 4204
5830 University Research Court
College Park, Maryland  20740
Phone: ‪(301) 683-1395‬

**Comment by zmoon** [comment]
> No warning. Compiled just fine

Great, do you want to approve then (seems I can't since I opened this PR, even though it's from your fork)

**Review by bbakernoaa — APPROVED** [review]

**Pull request merged**